js/content/store.js

```diff
@@ -3190,9 +3250,115 @@ let WishlistPageClass = (function(){
                 DynamicStore.clear();
                 location.reload();
             });
-        }, true)
+        }, true);
     }
 
+    WishlistPageClass.prototype.showModalDialog = function() {
+        ExtensionLayer.runInPageContext(`function() {
+            let options = {};
+            let Modal = ShowConfirmDialog(
+                "${Localization.str.export_wishlist}",
+                \`${this.exportModalTemplate}\`,
+                "${Localization.str.save}",
+                "${Localization.str.cancel}");
+
+            $J("#es_export_modal input[placeholder]").each(function() {
+                $J(this).attr('size', $J(this).attr('placeholder').length);
+            });
+
+            $J('#es_export_form').change(function() {
+                $J("#es_export_form").serializeArray().forEach((item) => options[item.name] = item.value);
+            });
+
+            Modal.done(function() {
+                let data = { rgAllApps: g_Wishlist.rgAllApps, g_rgAppInfo, options };
+                Messenger.postMessage("exportWishlist", data);
+            });
+        }`);
+
+        Messenger.addMessageListener("exportWishlist", (data) => {
+            let { rgAllApps, g_rgAppInfo, options } = data || {};
+            if (!rgAllApps || !g_rgAppInfo || !options) { return; }
+
+            let toexport, filename;
+            let notes = SyncedStorage.get("user_notes") || {};
+            let json = rgAllApps.map(appid => {
+                let appinfo = g_rgAppInfo[appid];
+                appinfo.appid = appid;
+                if (notes[appid]) {
+                    appinfo.note = notes[appid];
+                }
+                return appinfo;
+            });
+
+            if (options.export_type === "JSON") {
+                filename = "wishlist_export.json";
+                toexport = JSON.stringify(json, null, 4);
+                doExport(options.export_method, toexport, filename);
+                return;
+            }
+
+            let props = [];
+            let seps = [];
+            for (let key in options) {
+                if (key.includes("prop")) {
+                    props[+key.replace("prop", "")] = options[key];
+                }
+                if (key.includes("sep")) {
+                    seps[+key.replace("sep", "")] = options[key];
+                }
+            }
+
+            filename = "wishlist_export." + (options.export_type === "CSV" ? "csv" : "txt");
+            toexport = json.map(app => {
+                let line = "";
+                for (let i = 0; i < props.length; i++) {
+                    if (props[i]) {
+                        let prop = app[props[i]] || "";
+                        line += (options.export_type === "CSV" ? prop.replace(/,/g, "") : prop);
```

Comment:
This is not guaranteed to generate a valid CSV file and iirc `+=` is still the slowest way to build a string.
```js
var CSV = {
  escape(val) {
    if (typeof val == 'undefined' || val === null || val === '') {
      return '';
    }
    if (typeof val == 'string') {
      val = val.replace(/"/g, '""');
      val = `"${val}"`; // technically only required if val contains ',', '"' or "\n" but is correct even if it doesn't
    } else {
      val = `${val}`;
    }
    return val;
  },
  row(obj, props, sep=',') {
    return props.map(prop => CSV.escape(obj[prop])).join(sep);
  },
  file(collection, props, sep=',') {
    return [
      props.map(prop => CSV.escape(prop)).join(sep),
    ].concat(collection.map(obj => CSV.row(obj, props, sep))).join("\n");
  },
};
```

Reply:
I don't think this is necessary. In this case, `+=` is fine. Besides, this works well alongside the "text" format. I will implement the escaping, though.